lib/nimble_options.ex

```diff
@@ -440,6 +442,31 @@ defmodule NimbleOptions do
     end
   end
 
+  defp validate_type({:list, type}, key, values) when is_list(values) do
+    Enum.reduce_while(values, :ok, fn value, acc ->
+      case validate_type(type, key, value) do
+        :ok ->
+          {:cont, acc}
+
+        {:ok, _} ->
```

Comment:
We can't throw away the result here. For example, if you have this:

```elixir
{:list, {:custom, __MODULE__, :validate_port, []}}
```

and

```elixir
defp validate_port(string) when is_binary(string), do: {:ok, String.to_integer(string)}
defp validate_port(int) when is_integer(int), do: {:ok, int}
defp validate_port(other), do: {:error, "expected an integer or a string, got: #{inspect(other)}"}
```

then you will throw away the conversion from string to integer. We need to keep those values.

Reply:
@whatyouhide changed the approach on commit https://github.com/dashbitco/nimble_options/pull/35/commits/05b66ffb48e518aa57aee865b375b37d57ba56d0

When it's a custom, it will traverse the whole list to return the evaluated value (https://github.com/dashbitco/nimble_options/pull/35/commits/05b66ffb48e518aa57aee865b375b37d57ba56d0#diff-a948e6ad1d36e6ded3a51ec1fe9959acR468-R470) but for other types it'll stop as soon as an error is found and just return all values for the user (https://github.com/dashbitco/nimble_options/pull/35/commits/05b66ffb48e518aa57aee865b375b37d57ba56d0#diff-a948e6ad1d36e6ded3a51ec1fe9959acR478-R480), which I guess makes more sense for debugging.

The tricky part is how to deal with all error messages returned by a `:custom` validation (https://github.com/dashbitco/nimble_options/pull/35/commits/05b66ffb48e518aa57aee865b375b37d57ba56d0#diff-a948e6ad1d36e6ded3a51ec1fe9959acR494-R502) because there's no data structure to accumulate and format the message (like a changeset), so I'm just returning all evaluated error messages (https://github.com/dashbitco/nimble_options/pull/35/commits/05b66ffb48e518aa57aee865b375b37d57ba56d0#diff-6c7b866f243a12c3add997a69689b569R450)

Does that make sense?

Reply:
We can’t throw the result away here in any case, because we can have complex nested validations.

```elixir
    test "valid nested {:list, {:list, type}} for complex types" do
      schema = [ports: [type: {:list, {:list, {:custom, __MODULE__, :validate_port, []}}}]]

      opts = [ports: [[22], ["80", "8080"]]]
      assert {:ok, validated_opts} = NimbleOptions.validate(opts, schema)
      assert validated_opts[:ports] == [[22], [80, 8080]]
    end
```

```diff
  1) test type validation valid nested {:list, {:list, type}} for complex types (NimbleOptionsTest)
     test/nimble_options_test.exs:473
     Assertion with == failed
     code:  assert validated_opts[:ports] == [[22], [80, 8080]]
     - left:  [[22], ["80", "8080"]]
     + right: [[22], [80, 8080]]
     stacktrace:
       test/nimble_options_test.exs:478: (test)
```

Reply:
Hey @am-kantox just fixed nested list validation. Thanks.